aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
-    IntArrayRef batch_tensor1(tensor1.sizes().data(), std::max<int64_t>(dim_tensor1 - 2, 0));
-    int64_t m2 = dim_tensor2 > 1 ? tensor2.size(-2) : 1;
-    int64_t p = tensor2.size(-1);
-    IntArrayRef batch_tensor2(tensor2.sizes().data(), std::max<int64_t>(dim_tensor2 - 2, 0));
-
-    // expand the batch portion (i.e. cut off matrix dimensions and expand rest)
-    DimVector expand_batch_portion = infer_size_dimvector(batch_tensor1, batch_tensor2);
-
-    DimVector tensor1_expand_size(expand_batch_portion);
-    tensor1_expand_size.push_back(n);
-    tensor1_expand_size.push_back(m1);
-
-    DimVector tensor2_expand_size(expand_batch_portion);
-    tensor2_expand_size.push_back(m2);
-    tensor2_expand_size.push_back(p);
-
-    const int64_t expand_batch_product =
-        c10::multiply_integers(expand_batch_portion);
-
-    std::array<int64_t, 3> tensor1_bmm_view = {expand_batch_product, n, m1};
-
-    std::array<int64_t, 3> tensor2_bmm_view = {expand_batch_product, m2, p};
+    const int64_t n = dim_tensor1 > 1 ? tensor1.sizes().cend()[-2] : 1LL;
+    const int64_t m1 = tensor1.sizes().back();
+    const IntArrayRef batch_tensor1(tensor1.sizes().data(),
+                                    std::max<int64_t>(dim_tensor1 - 2, 0LL));
+    const int64_t m2 = dim_tensor2 > 1 ? tensor2.sizes().cend()[-2] : tensor2.sizes().back();
+    const int64_t p = dim_tensor2 > 1 ? tensor2.sizes().back() : 1LL;
+    const IntArrayRef batch_tensor2(tensor2.sizes().data(),
+                                    std::max<int64_t>(dim_tensor2 - 2, 0LL));
+    auto output_shape = infer_size_dimvector(batch_tensor1, batch_tensor2);
+
+    const auto tensor1_expand_size = [&output_shape, n, m1]{ DimVector ret(output_shape);
+                                                             ret.append({n, m1});
+                                                             return ret; }();
+    const auto tensor2_expand_size = [&output_shape, m2, p]{ DimVector ret(output_shape);
+                                                             ret.append({m2, p});
+                                                             return ret; }();
+
+    const int64_t expand_batch_product = c10::multiply_integers(output_shape);
 
     // flatten expanded batches
-    Tensor tensor1_expanded = tensor1.expand(tensor1_expand_size).reshape(tensor1_bmm_view);
-    Tensor tensor2_expanded = tensor2.expand(tensor2_expand_size).reshape(tensor2_bmm_view);
-
-    // reshape batches back into result
-    DimVector output_shape(expand_batch_portion);
+    const auto tensor1_expanded = tensor1.expand(tensor1_expand_size)
+                                         .reshape({expand_batch_product, n, m1});
+    const auto tensor2_expanded = tensor2.expand(tensor2_expand_size)
+                                         .reshape({expand_batch_product, m2, p});
     if (dim_tensor1 > 1) {
       output_shape.push_back(n);
     }
     if (dim_tensor2 > 1) {
       output_shape.push_back(p);
     }
```

Comment:
These cases are always true now, no?

Reply:
Actually maybe this is not true but then the comment at the top of this block is misleading

Reply:
I guess you refer to the comment
>     // We are multiplying b1 x n x m1 by b2 x m2 x p (where b1 and b2 can be lists);

That's fair, I'll edit that one.